in_toto/keylib.go

```diff
@@ -2,220 +2,330 @@ package in_toto
+LoadKey loads the key file at specified file path into the key object.
+It automatically derives the PEM type and the key type.
+Right now the following PEM types are supported:
+
+	* PKCS1 for private keys
+	* PKCS8	for private keys
+	* PKIX for public keys
+
+The following key types are supported:
 
-	var keyObj Key
-	err := json.Unmarshal([]uint8(JSONString), &keyObj)
+	* ed25519
+	* RSA
+
+On success it will return nil. The following errors can happen:
+
+	* path not found or not readable
+	* no PEM block in the loaded file
+	* no valid PKCS8/PKCS1 private key or PKIX public key
+	* errors while marshalling
+	* unsupported key types
+*/
+func (k *Key) LoadKey(path string, scheme string, keyIdHashAlgorithms []string) error {
+	pemFile, err := os.Open(path)
 	if err != nil {
-		return keyObj, fmt.Errorf("this is not a valid JSON key object")
+		return err
 	}
-
-	if keyObj.KeyType != "ed25519" || keyObj.Scheme != "ed25519" {
-		return keyObj, fmt.Errorf("this doesn't appear to be an ed25519 key")
+	defer func() {
+		if closeErr := pemFile.Close(); closeErr != nil {
+			err = closeErr
+		}
+	}()
+	// Read key bytes and decode PEM
+	pemBytes, err := ioutil.ReadAll(pemFile)
+	if err != nil {
+		return err
 	}
 
-	if keyObj.KeyVal.Private == "" {
-		return keyObj, fmt.Errorf("this key is not a private key")
+	// TODO: There could be more key data in _, which we silently ignore here.
+	// Should we handle it / fail / say something about it?
+	data, _ := pem.Decode(pemBytes)
+	if data == nil {
+		return ErrNoPEMBLock
 	}
 
-	// 64 hexadecimal digits => 32 bytes for the private portion of the key
-	if len(keyObj.KeyVal.Private) != 64 {
-		return keyObj, fmt.Errorf("the private field on this key is malformed")
+	// Try to load private key, if this fails try to load
+	// key as public key
+	key, err := ParseKey(data.Bytes)
+	if err != nil {
+		return err
 	}
 
-	return keyObj, nil
+	// Use type switch to identify the key format
+	switch key.(type) {
+	case *rsa.PublicKey:
+		if err := k.SetKeyComponents(pemBytes, []byte{}, "rsa", scheme, keyIdHashAlgorithms); err != nil {
+			return err
+		}
+	case *rsa.PrivateKey:
+		// Note: We store the public key as PKCS8 key here, although the private key get's stored as PKCS1 key
+		// This behavior is consistent to the securesystemslib
+		pubKeyBytes, err := x509.MarshalPKIXPublicKey(key.(*rsa.PrivateKey).Public())
+		if err != nil {
+			return err
+		}
+		if err := k.SetKeyComponents(pubKeyBytes, pemBytes, "rsa", scheme, keyIdHashAlgorithms); err != nil {
+			return err
+		}
+	case ed25519.PublicKey:
+		if err := k.SetKeyComponents(key.(ed25519.PublicKey), []byte{}, "ed25519", scheme, keyIdHashAlgorithms); err != nil {
+			return err
+		}
+	case ed25519.PrivateKey:
+		pubKeyBytes := key.(ed25519.PrivateKey).Public()
+		if err := k.SetKeyComponents(pubKeyBytes.(ed25519.PublicKey), key.(ed25519.PrivateKey), "ed25519", scheme, keyIdHashAlgorithms); err != nil {
+			return err
+		}
```

Comment:
same here, could we set the parameters and then call "set key components" afterwards? It'd be easier to grok what the difference between these cases are.

Reply:
Actually i tried to make it this way, but due to Go's type system it's rather difficult.
I tried something like this:
```
case ed25519.PublicKey:
    key = key.(ed25519.PublicKey)
```

and then just setting the struct on the end with the set parameters, but Go's type system says "no no" :> because I would change the type of key from interface to ed25519.PublicKey. At the same time, functions like x509.MarshalPKIXPublicKey need a specific key type and can't work on an interface, except we specify the type.